src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1214,9 +1236,56 @@ public bool StatusCanHaveBody(int statusCode)
                    statusCode != 304;
         }
 
-        private void ThrowResponseAlreadyStartedException(string value)
+        private void RejectNonBodyTransferEncodingResponse(bool appCompleted)
+        {
+            var ex = BadHttpResponseException.GetException(ResponseRejectionReason.NonBodyResponseTransferEncodingSet, StatusCode);
+            if (!appCompleted)
+            {
+                // Back out of header creation surface exeception in user code
+                _requestProcessingStatus = RequestProcessingStatus.RequestStarted;
```

Comment:
If we're logging an not throwing for HEAD requests in `HandleNonBodyResponseWrite`, shouldn't we just remove the header and do the same here? I think we should be consistent.